common/numerated/src/numerated.rs

```diff
@@ -0,0 +1,196 @@
+// This file is part of Gear.
+
+// Copyright (C) 2023 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! [Numerated], [Bound] traits definition and implementation for integer types.
+
+use num_traits::{bounds::UpperBounded, One, PrimInt, Unsigned};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum BoundValue<T: Sized> {
+    /// The bound is a value.
+    Value(T),
+    /// The bound is an upper bound. Contains `T` max value.
+    Upper(T),
+}
+
+/// For any type `T`, `Bound<T>` is a type, which has set of values bigger than `T` by one element.
+/// - Each value from `T` has unambiguous mapping to `Bound<T>`.
+/// - Each value from `Bound<T>`, except one called __upper__, has unambiguous mapping to `T`.
+/// - __upper__ value has no mapping to `T`, but can be used to get `T` max value.
+///
+/// # Examples
+/// 1) For any `T`, which max value can be get by calling some static live time function,
+/// Option<T> can be used as `Bound<T>`. `None` is __upper__. Mapping: Some(t) -> t, t -> Some(t).
```

Comment:
```suggestion
/// `Option<T>` can be used as `Bound<T>`. `None` is __upper__. Mapping: Some(t) -> t, t -> Some(t).
```

Reply:
We leave possibility for user to choose bound

Reply:
this emits warning while building docs ))

Reply:
Lol, ok)) does not understand correctlly. Fixed